### DISCUSSION TITLE
Allow integrations to use the Scheduler

### DISCUSSION THREAD

**Discussion by IgnacioHR** [home-assistant/architecture#876] [Entity Models]
### Problem

The current Scheduler (helper) can be used to create schedules to control on/off status of another device. This works great when HA (core) takes full control of the target device and when the device has no an internal scheduler. For devices that already have an scheduler, the current scheduler (helper) can't be used to edit the device schedule.

### Proposal

Create a new Schedule (entity) that will be edited using the Schedule (helper) and allow integrations to provide its own versions of the Schedule (entity) so the integration can provide scheduler from the device and store back the user changes to the device.

The steps might be:
1. Rename current `scheduler` helper as `input_scheduler`
2. Create new `scheduler` entity that represents the internally stored schedules and can be used by integrations

The idea is that all existing schedule info stored will remain working as is (smooth integration) but the new schedule entity is available to be used by any other integration.

### A bit of history

About a year ago I stared creating an integration for my boiler at home, the boiler is a De Dietrich C230 ECO. The boiler provides an RS485 port with MODBUS protocol and I already found the information about stored parameters. The schedule for the 5 circuits is available as 42 bytes per week (every bit represents an interval of 30 minutes). I managed to create Schedules in yaml to represent the internal boiler schedule using the Schedule (helper) and it works! But it is a bit of a hack as there is no a clean way to do it because of lack of `Platform.SCHEDULE` integration. During the development of my integration, I notices about the limitations of the current scheduler (helper) and I think I've a clean idea about how it should be implemented.

### Who might be interested

Of course, I'm interested, but there are plenty of devices out there that already provides an internal scheduler. This is just a short list of them

* Heating boilers
* Vaccum cleaners
* Security panels
* Irrigation programmers
* Pool treatment plants
* ...

### Who will spend time on this?

Of course, I'll do the job! I can't promise to just do the job in two days because I've other things to do in my real live and my time to spend on this is limited, but it is a year now since I started with my integration and I feel I can do it.

I already finished step 1 and made about 80% progress of step 2 but all testing and documentation is still missing.

### References

[Right way of using scheduler from third party integrations](https://community.home-assistant.io/t/right-way-of-using-scheduler-from-third-party-integrations/465478)
[WTH can’t use schedule UI to edit a schedule in my custom integration](https://community.home-assistant.io/t/wth-cant-use-schedule-ui-to-edit-a-schedule-in-my-custom-integration/471873)

**Comment by frenck** [comment]
Which other core integrations can directly implement this new entity type? We don't want to add entity components for a single-use case (in this case just Modbus). If we implement/add entity components, there should be (or at least have the existing potential to be) widely used in our core codebase.

The proposal is currently based on solving a specific modbus use-case and not Home Assistant as a whole.

../Frenck

**Comment by IgnacioHR** [comment]
Hmm.. Maybe I've not explained properly the use case. MODBUS is not related with the schedule (entity) I'm proposing to do.

I fully agree with you, a schedule (entity) must be a generic component used by any other integration regardless of the protocol to send the updates to the target device. In fact, the current scheduler (helper) can't be used by integrations because it only stores to yaml internally.

In my particular use case, in my own integration, (out of this architectural discussion) I'll extend the generic schedule (entity) to write to http because I'm using http to talk to talk to a "black box" that converts to MODBUS. Other integrations will have to do the same.

**Reply by frenck** [reply]
> Hmm.. Maybe I've not explained properly the use case. MODBUS is not related with the schedule (entity) I'm proposing to do.

Sure, but that was the only example given, we need to have a sustainable possible base/list of candidates before we introduce something like this, which is missing in your proposal.

> Other integrations will have to do the same.

ok, but... the architectural proposal is missing the base of integrations in our core integrations that could implement this today? As in, if this would be accepted, we need to know this is a common use case that will be leverages by X (number) of integrations. If there are no other integrations that will implement this, then we have no use in adding the proposal.

Recap: Which core integrations would be able to use the proposal? Could you provide a list of integrations that would provide a schedule?

**Reply by IgnacioHR** [reply]
Here is a quick research, note, I ask ChatGPT directly!

Here is a list of some home automation devices with timers that have integration with Home Assistant:

Smart vacuums: iRobot Roomba, Neato Botvac, Ecovacs Deebot, Roborock S5, and Xiaomi Mi Robot Vacuum.

Smart lights: Philips Hue, LIFX, Sengled, Yeelight, and TP-Link Kasa.

Smart thermostats: Nest, Ecobee, Honeywell, Tado, and Netatmo.

Smart locks: August Smart Lock, Schlage Encode, Yale Real Living, Kwikset SmartCode, and Danalock.

Smart plugs: TP-Link Kasa, Belkin WeMo, iHome, Aeotec Smart Switch, and Sonoff.

Smart sensors: Aqara, Fibaro, SmartThings, Xiaomi, and Aeotec.

All of these devices have the ability to schedule tasks and automated routines, making them great for integration with Home Assistant. Additionally, many of these devices can be controlled through mobile apps and voice assistants such as Google Assistant and Amazon Alexa.

For more information about integrating these devices with Home Assistant, you can check out the official Home Assistant documentation or search online for specific tutorials for each device.

And Rain Bird irrigation! (edited)

**Reply by frenck** [reply]
Are you serious?

https://www.home-assistant.io/blog/2023/01/23/help-others-leave-ai-at-the-dor/

**Reply by IgnacioHR** [reply]
I'm sorry it's first time I see that link. But the response is not that wrong at all. A scheduler is a very common component we use every day to automate things.

**Reply by frenck** [reply]
Is it? Could you point me to technical details on each of the mentioned integrations above, on where their APIs implement and expose scheduling that Home Assistant could wrap into a schedule entity?

../Frenck

**Reply by IgnacioHR** [reply]
No, I can't. My request is the reaction of my initial frustration trying to do it for my own integration after discovering the lack of base functionality (lack of a scheduler entity) that is (maybe) the root cause for no one else trying to provide a scheduler along with their integration even if the device provides an internal scheduler.

Rain Bird is one example I've one at home to control irrigation of my garden and I've to go manually and use their hardware to manage the scheduler.  https://www.home-assistant.io/integrations/rainbird/

iRobot Roomba and Braava - https://www.home-assistant.io/integrations/roomba/

My De Dietrich boiler is another example. 

There is a long discussion here (https://community.home-assistant.io/t/scheduler-card-custom-component/217458/1844) of people trying to implement schedulers using yaml and other efforts

To mention just a few in the last months

https://community.home-assistant.io/t/scheduler-card-custom-component/217458/1846
https://community.home-assistant.io/t/scheduler-card-custom-component/217458/1851
https://community.home-assistant.io/t/scheduler-card-custom-component/217458/1811

**Reply by frenck** [reply]
Just a note, as it was referenced in your last comment a lot: Custom integrations are not a consideration when implementing base entity components. Those thus have no weight in this proposal.

> No, I can't.

In that case, this proposal will be declined. We need to have a proven and solid base for an entity component to have a general use. Without researching this, it won't be possible to say it has that or not.

>  the root cause for no one else trying to provide a scheduler along with their integration even if the device provides an internal scheduler.

That is an assumption. Let's make it factual, through research.

> Rain Bird is one example I've one at home to control irrigation of my garden and I've to go manually and use their hardware to manage the scheduler.

Right, but many things have schedulers, the question here is: Can they be used by HA, is the API exposed? And if the API was available, and would it fit the scheduler implementation/architecture we have? Are there differences? If so, what do we do to adopt to it? 

This same search and questions need to be repeated and refined for e.g., Roomba and other potential integrations.

Only when there is a wider possible use-case (that is actually implementable) across our core integrations, new entity components can be considered.

../Frenck

**Reply by allenporter** [reply]
We've closed this issue, and it should remain closed for now, but i wanted to add one more thought here since we called out rainbird specifically. Rainbird does have a schedule API. So far the "read" part is working in a [calendar](https://github.com/home-assistant/core/pull/87604 ) and the "write" uses a similar API not yet added to the python library, but it could be ([commands](https://github.com/allenporter/pyrainbird/blob/main/pyrainbird/resources/sipcommands.yaml#L64).

So, Frenck's questions are the right ones. For this example, I would say that yes the API is exposed and usable, but the API differences would be many.  When you actually start looking into the details, it gets quite complicated quickly, so even if we do have supporting APIs from a bunch of integrations we'll probably see they have very different requirements that are difficult to unify.  For example if we look at the rainbird [parameters](https://github.com/allenporter/pyrainbird/blob/bc7502130e0ee6d1092c6f3d373d6d8a3bbf46f5/pyrainbird/data.py#L441) there are:
- Program of "odd days", "event days", "cycle", "custom"
- In cycle mode: which days to run on
- In custom mode: the set of days of the week
- A list of start times, for the specified days 
- A list of zones affected
- A list of durations for each zone

These various flavors would be difficult to translate back and forth between a schedule entity and represent in the home assistant UI as a schedule. Since it would only probably make sense to have home assistant support a subset of features, then it wouldn't really fully have parity with the native app and wouldn't work well.  Then you also have to do with limits (e.g. it only supports 4 start times per day).  This is why i ended up just representing the above as a read-only calendar.

**Comment by frenck** [comment]
This proposal is missing actual references of API and integrations that could implement this usecase today.

Going to friendly decline and close this proposal for now for that reason.

../Frenck
